src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/simple/RandomizedJoinPlanOptimizerImpl.java

```diff
@@ -0,0 +1,50 @@
+package se.liu.ida.hefquin.engine.queryproc.impl.optimizer.simple;
+
+import java.util.List;
+import java.util.Random;
+
+import se.liu.ida.hefquin.engine.queryplan.PhysicalPlan;
+import se.liu.ida.hefquin.engine.queryplan.utils.PhysicalPlanFactory;
+import se.liu.ida.hefquin.engine.queryproc.QueryOptimizationException;
+
+public class RandomizedJoinPlanOptimizerImpl implements JoinPlanOptimizer {
+	
+	/**
+	 * The class contains a random seed object.
+	 * This cannot be marked final (compilation error if so).
+	 */
+	protected Random random = new Random();

```

Comment:
```suggestion
	 */
	protected final Random random = new Random();
```

You can make this member variable `final`. Your mistake was that, after assigning something to it here, you tried to reassign something else again in the constructor. That's not possible with `final`. In contrast, every variable marked as `final` can be assigned only once---that's the meaning of the `final` keyword in Java.

Notice that this is different from `const` in C++.  `final` in Java is only about the variable, not about the object assigned to the variable. Hence, it is still possible to call member functions that may change the internal state of the object.